lib/ace/edit_session/bracket_match.js

```diff
@@ -37,10 +37,71 @@ var Range = require("../range").Range;
 
 function BracketMatch() {
 
+
+    this.findSurroundingBrackets = function(position) {
+        var allBrackets = { 
+            left : [this.$findOpeningBracket("}", position, /(\.?.paren)+/), 
+                    this.$findOpeningBracket(']', position, /(\.?.paren)+/), 
+                    this.$findOpeningBracket(')', position, /(\.?.paren)+/)],
+            right : [this.$findClosingBracket("{", position, /(\.?.paren)+/),
+                     this.$findClosingBracket('[', position, /(\.?.paren)+/),
+                     this.$findClosingBracket('(', position, /(\.?.paren)+/)]
+        };
+
+        var leftNearest = null,
+            rightNearest = null;
+
+
+        for (var key = 0; key < allBrackets.left.length; key++) {
+
+            var leftCandidate = allBrackets.left[key],
+                rightCandidate = allBrackets.right[key];
+
+            if (!leftNearest)
+                leftNearest = leftCandidate;
+            if (!rightNearest)
+                rightNearest = rightCandidate;
+ 
+            if(leftCandidate)
+                if (leftNearest.row <= leftCandidate.row)
+                    if (leftNearest.row == leftCandidate.row) {
+                        if (leftNearest.column < leftCandidate.column)
+                            leftNearest = leftCandidate;
+                    } else {
+                        leftNearest = leftCandidate;
+                    }
+
+            if (rightCandidate)
+                if (rightNearest.row >= rightCandidate.row)
+                    if (rightNearest.row == rightCandidate.row) {
+                        if (rightNearest.column > rightCandidate.column)
+                            rightNearest = rightCandidate;
+                    } else { 
+                        rightNearest = allBrackets.rights[key];
+                    }
+        }
+
+        if (!leftNearest || !rightNearest) {
```

Comment:
i think that this will be more readable if you create an object:

```
var result = {
  left: leftNearest,
  right: rightNearest,
  mismatch: true
}
```

and then update mismatch field using consition from line 92.